farm-bot/codebase/contracts/farm-bot.sol

```diff
@@ -2,62 +2,205 @@
+        bool transferSuccess = stakingToken.transfer(msg.sender, _lpAmount);
         require(transferSuccess, "Transfer failed, aborting withdrawal");
-        fpBalances[msg.sender] -= _fpAmount;
-        fpTotalBalance -= _fpAmount;
+	_burn(msg.sender, _fpAmount);
         lpTotalBalance -= _lpAmount;
     }
 
-    function investInFarm() public {
-        // todo eventually make private. The public function will do more: reap rewards, swap them, obtain LP tokens, and then this.
-        uint256 tokenBalance = IERC20(tokenAddress).balanceOf(address(this));
+    function investInFarm() private {
+        uint256 tokenBalance = stakingToken.balanceOf(address(this));
         require(tokenBalance > 0, "Cannot invest in farm because tokenBalance is 0");
-        IERC20(tokenAddress).approve(address(farm), tokenBalance);
-        farm.stake(tokenBalance);
+        stakingToken.approve(address(stakingRewards), tokenBalance);
+        stakingRewards.stake(tokenBalance);
+    }
+
+    function claimRewards(uint deadline) public ensure(deadline) {
+        // todo eventually make private
+        stakingRewards.getReward();
+	uint256 tokenBalance = rewardsToken.balanceOf(address(this));
+
+	if (tokenBalance == 0) {
+	    return;
+	}
+
+	uint256 feeAmount = tokenBalance * feeNumerator / feeDenominator;
+	uint256 halfTokens = (tokenBalance - feeAmount) / 2;
+
+	uint256 amountToken0;
+	// Figure out best-case scenario amount of token0 we can get and swap
+	if (path0.length >= 2) {
+	    uint[] memory amountsOut = router.getAmountsOut(halfTokens, path0);
+	    uint amountOut = amountsOut[amountsOut.length-1];
+	    rewardsToken.approve(address(router), halfTokens);
+	    uint[] memory amountsSwapped = router.swapExactTokensForTokens(
+                halfTokens,
+		amountOut * slippageNumerator / slippageDenominator,
+		path0,
+		address(this),
+		deadline
+            );
+	    amountToken0 = amountsSwapped[amountsSwapped.length-1];
+	} else {
+	    amountToken0 = halfTokens;
+	}
+
+	uint256 amountToken1;
+	// Figure out best-case scenario amount of token1 we can get and swap
+	if (path1.length >= 2) {
+	    uint[] memory amountsOut = router.getAmountsOut(halfTokens, path1);
+	    uint amountOut = amountsOut[amountsOut.length-1];
+	    rewardsToken.approve(address(router), halfTokens);
+	    uint[] memory amountsSwapped = router.swapExactTokensForTokens(
+                halfTokens,
+		amountOut * slippageNumerator / slippageDenominator,
+		path1,
+		address(this),
+		deadline
+            );
+	    amountToken1 = amountsSwapped[amountsSwapped.length-1];
+	} else {
+	    amountToken1 = halfTokens;
+	}
+
+	// Approve the router to spend the bot's token0/token1
+	stakingToken0.approve(address(router), amountToken0);
+	stakingToken1.approve(address(router), amountToken1);
+	// Actually add liquidity
+	router.addLiquidity(
+	    address(stakingToken0),
+	    address(stakingToken1),
+	    amountToken0,
+	    amountToken1,
```

Comment:
```suggestion
Not important for the MVP, but I wonder if this would help us collect dust:
```suggestion
	    stakingToken0.balanceOf(address(this)),
	    stakingToken1.balanceOf(address(this)),
```

Reply:
that way we at least try to dump the leftovers from slippage in the past